### ISSUE TITLE
Client side cert validation by default

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#15216]
Should `SSL_VERIFY_PEER` to verify the server's certificate be the default on the client?

Also should we fail by default if `SSL_set1_host()` has not been called on the client and `SSL_VERIFY_PEER` is set?

Ideally we should fail "safe". The current defaults don't do that.

If the defaults are changed then we would have to be careful not to fail if custom `X509_VERIFY_PARAM`s are in use, or a custom verify cert callback is in use. Applications that do that may be verifying the hostname in other ways. But my expectation is that the majority of applications probably don't do that.

**Comment by mattcaswell** [comment]
Ping @vdukhovni

**Comment by richsalz** [comment]
A further complication is where the client-side looks for the server name. (For example, using the CN field is a deprecated for more than ten years.)

I think is `VERIFY_PEER` becomes the default, and the client doesn't specify the hostname, then you just look at the chain.

**Comment by mattcaswell** [comment]
> then you just look at the chain.

You mean default to not verifying the hostname? This doesn't seem quite in the spirit of secure by default.

**Comment by richsalz** [comment]
If you are going to enable the verify flag, and the client does not specify a hostname, that seems the least disruptive. On the other hand, we're turning off various renegotiation options, so if you disagree with that view it's not a big deal to me.

**Comment by mattcaswell** [comment]
It *is* likely to be a disruptive change if we decide to do it. But if we're willing to take the hit I think it would bring great security benefits to our users.

**Comment by richsalz** [comment]
I agree the benefits outweigh the concerns, and I think setting the hostname meaning "any host is okay" is a good tradeoff as I said I don't feel strongly to push very hard.  An alternative of having a new unique error code is also good.

**Comment by mattcaswell** [comment]
OTC: Moved to post 3.0 milestone

**Comment by mattcaswell** [comment]
OTC: Move to 4.0 milestone

**Comment by nhorman** [comment]
@mattcaswell this has been moved to subsequent major releases twice now.  Is this realistically something that we can plan to do?  If so, should we prioritize it for 4.0?  If so, I could really use some input on a more fine grained set of details involved in implementing it.

**Comment by nhorman** [comment]
marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by t8m** [comment]
@nhorman This was moved just once - from 3.0 to 4.0.

There are a couple of OTC hold issues with the 4.0 milestone. We do not look at them currently but we should look at them before we start planning for the 4.0 release.

**Comment by vdukhovni** [comment]
Note that requiring hostname checks by default could break typical DANE-EE(3) use-cases or equivalently RPK authentication (which is based on synthetic DANE-EE(3) records).

So any such change needs to happen with care, because there are other ways in SSL sessions can be authenticated. There are also various PSK modes, where certificates aren't even exchanged, or are tangential to peer authentication.

And of course extant unauthenticated opportunistic TLS use-cases. So at the very least:

* Explicitly selecting SSL_VERIFY_NONE in the SSL_CTX or SSL handle must continue to work without name checks.
* DANE-EE(3) must be able to function with no hostname checks with `DANE_FLAG_NO_DANE_EE_NAMECHECKS` in effect.

**Comment by t8m** [comment]
This should be moved to a community discussions. Closing here.
